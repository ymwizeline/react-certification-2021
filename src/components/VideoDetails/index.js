export { default } from './VideoDetails.component';
