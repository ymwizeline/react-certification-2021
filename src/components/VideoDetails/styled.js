import styled from 'styled-components';

export const Container = styled.div`
  display: flex;
  flex-direction: column;
`;

export const Button = styled.button`
  width: 150px;
  height: 40px;
`;
