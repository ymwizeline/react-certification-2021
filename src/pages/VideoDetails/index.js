export { default } from './VideoDetails.page';
