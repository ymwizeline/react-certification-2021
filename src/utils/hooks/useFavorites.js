import { useState, useEffect } from 'react';
import { storage } from '../storage';
import { FAVORITES_STORAGE_KEY } from '../constants';

export const useFavorites = (isFavorite) => {
  const [favorites, setFavorites] = useState(null);

  useEffect(() => {
    const localFavorites = storage.get(FAVORITES_STORAGE_KEY);
    if (localFavorites) setFavorites(localFavorites);
  }, [isFavorite]);

  return favorites;
};
