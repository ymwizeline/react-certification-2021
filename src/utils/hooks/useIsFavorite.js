import { useState, useEffect } from 'react';
import { FAVORITES_STORAGE_KEY } from '../constants';
import { storage } from '../storage';

export const useIsFavorite = (videoId) => {
  const [isFavorite, setIsFavorite] = useState(false);

  useEffect(() => {
    const favorites = storage.get(FAVORITES_STORAGE_KEY);
    if (favorites && favorites[videoId]) setIsFavorite(true);
  }, [videoId]);

  return { isFavorite, setIsFavorite };
};
