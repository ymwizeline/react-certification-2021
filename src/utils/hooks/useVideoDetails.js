import { useContext, useEffect, useState } from 'react';
import { useHistory } from 'react-router-dom';
import { Context } from '../../context';
import { storage } from '../storage';
import { FAVORITES_STORAGE_KEY } from '../constants';

export const useVideoDetails = (id) => {
  const {
    state: { videos },
  } = useContext(Context);
  const history = useHistory();
  const [videoDetails, setVideoDetails] = useState(null);
  const [similar, setSimilar] = useState(null);

  useEffect(() => {
    const data = videos.filter(({ id: { videoId } }) => id === videoId);
    if (!data.length) {
      const favorites = storage.get(FAVORITES_STORAGE_KEY);
      if (favorites && favorites[id]) setVideoDetails(favorites[id]);
      else history.push(`/`);
    } else setVideoDetails(data[0]);
  }, [videos, id, history]);

  useEffect(() => {
    const data = videos.filter(({ id: { videoId } }) => id !== videoId);
    setSimilar(data);
  }, [videos, id]);

  return { videoDetails, similar };
};
