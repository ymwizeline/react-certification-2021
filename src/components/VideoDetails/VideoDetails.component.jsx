import React, { useContext } from 'react';
import { Context } from '../../context';
import VideoPlayer from '../VideoPlayer';
import SimilarVideo from '../SimilarVideo';
import { useVideoDetails } from '../../utils/hooks/useVideoDetails';
import { Container, PlayerContainer, SimilarContainer, Title, Button } from './styled';

const VideoDetails = ({ id }) => {
  const { videoDetails, similar } = useVideoDetails(id);
  const { dispatch } = useContext(Context);

  const link = `https://www.youtube.com/embed/${id}`;

  const goBack = () => {
    dispatch({
      type: 'CLOSE_VIDEO',
    });
  };
  return (
    <Container>
      <PlayerContainer>
        {videoDetails ? (
          <>
            <VideoPlayer src={link} title={videoDetails.snippet.title} />
            <Title>{videoDetails.snippet.title}</Title>
          </>
        ) : (
          <p>Loading player...</p>
        )}

        <Button type="button" onClick={goBack}>
          Go back
        </Button>
      </PlayerContainer>

      <SimilarContainer>
        {similar && similar.map((item) => <SimilarVideo key={item.etag} data={item} />)}
      </SimilarContainer>
    </Container>
  );
};

export default VideoDetails;
