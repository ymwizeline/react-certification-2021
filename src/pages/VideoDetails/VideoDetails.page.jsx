import React from 'react';
import { useHistory } from 'react-router-dom';
import { useAuth } from '../../providers/Auth';
import VideoPlayer from '../../components/VideoPlayer';
import SimilarVideo from '../../components/SimilarVideo';
import { useVideoDetails } from '../../utils/hooks/useVideoDetails';
import { useIsFavorite } from '../../utils/hooks/useIsFavorite';
import { useFavorites } from '../../utils/hooks/useFavorites';
import { Section, Title } from '../Home/styled';
import { storage } from '../../utils/storage';
import { FAVORITES_STORAGE_KEY } from '../../utils/constants';
import {
  Container,
  PlayerContainer,
  SimilarContainer,
  VideoTitle,
  Button,
} from './styled';

const VideoDetails = ({
  match: {
    params: { id, favorite },
  },
}) => {
  const { videoDetails, similar } = useVideoDetails(id);
  const { isFavorite, setIsFavorite } = useIsFavorite(id);
  const favorites = useFavorites(isFavorite);
  const { authenticated } = useAuth();
  const history = useHistory();

  const link = `https://www.youtube.com/embed/${id}`;

  const goBack = () => history.push('/');

  const addToFavorites = () => {
    const localFavorites = storage.get(FAVORITES_STORAGE_KEY);
    const updatedFavorites = { ...localFavorites };
    updatedFavorites[id] = videoDetails;
    storage.set(FAVORITES_STORAGE_KEY, updatedFavorites);
    setIsFavorite(true);
  };

  const removeFromFavorites = () => {
    const localFavorites = storage.get(FAVORITES_STORAGE_KEY);
    const updatedFavorites = { ...localFavorites };
    updatedFavorites[id] = undefined;
    storage.set(FAVORITES_STORAGE_KEY, updatedFavorites);
    setIsFavorite(false);
  };

  return (
    <Section className="homepage">
      <Title>{isFavorite ? 'Favorite details' : 'Video details'}</Title>
      <Container>
        <PlayerContainer>
          {videoDetails ? (
            <>
              <VideoPlayer src={link} title={videoDetails.snippet.title} />
              <VideoTitle>{videoDetails.snippet.title}</VideoTitle>
            </>
          ) : (
            <p>Loading player...</p>
          )}

          {authenticated && isFavorite && (
            <Button type="button" onClick={removeFromFavorites}>
              Remove from favorites
            </Button>
          )}
          {authenticated && !isFavorite && (
            <Button type="button" onClick={addToFavorites}>
              Add to favorites
            </Button>
          )}
          <Button type="button" onClick={goBack}>
            Go to Homepage
          </Button>
        </PlayerContainer>

        <SimilarContainer>
          {favorites &&
            favorite &&
            Object.values(favorites)
              .filter(({ id: { videoId } }) => videoId !== id)
              .map((item) => <SimilarVideo key={item.etag} data={item} />)}
          {similar &&
            !favorite &&
            similar.map((item) => <SimilarVideo key={item.etag} data={item} />)}
        </SimilarContainer>
      </Container>
    </Section>
  );
};

export default VideoDetails;
