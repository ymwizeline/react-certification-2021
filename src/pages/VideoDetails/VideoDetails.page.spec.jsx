import React from 'react';
import { render } from '@testing-library/react';
import VideoDetails from './VideoDetails.page';
import ContextProvider from '../../context';
import AuthProvider from '../../providers/Auth';
import { useVideoDetails } from '../../utils/hooks/useVideoDetails';
import mockData from '../../jsons/youtube-videos-mock.json';

const mockItem = mockData.items[1];

jest.mock('../../utils/hooks/useVideoDetails');
useVideoDetails.mockImplementation(() => ({
  videoDetails: mockItem,
}));

const props = { match: { params: { id: mockItem.id.videoId } } };

describe('VideoDetails component', () => {
  it('renders properly', () => {
    const { container } = render(
      <ContextProvider>
        <AuthProvider>
          <VideoDetails {...props} />
        </AuthProvider>
      </ContextProvider>
    );
    expect(container).toMatchSnapshot();
  });
});
