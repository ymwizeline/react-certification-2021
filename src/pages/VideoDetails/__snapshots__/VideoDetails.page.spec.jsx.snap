// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VideoDetails component renders properly 1`] = `
<div>
  <section
    class="sc-jSgupP gMkJrr homepage"
  >
    <h1
      class="sc-gKsewC jwTLYb"
    >
      Video details
    </h1>
    <div
      class="sc-iBPRYJ kSJANn"
    >
      <div
        class="sc-fubCfw ccLhuW"
      >
        <div
          class="sc-bdfBwQ kKsUrJ"
        >
          <iframe
            allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture"
            allowfullscreen=""
            class="sc-gsTCUz jINteo"
            frameborder="0"
            src="https://www.youtube.com/embed/nmXMgqjQzls"
            title="Video Tour | Welcome to Wizeline Guadalajara"
          />
        </div>
        <h2
          class="sc-jrAGrp dMKXCW"
        >
          Video Tour | Welcome to Wizeline Guadalajara
        </h2>
        <button
          class="sc-kEjbxe gNCMqM"
          type="button"
        >
          Go to Homepage
        </button>
      </div>
      <div
        class="sc-pFZIQ dyfyFe"
      />
    </div>
  </section>
</div>
`;
